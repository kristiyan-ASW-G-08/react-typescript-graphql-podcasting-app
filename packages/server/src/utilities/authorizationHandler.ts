import { AuthenticationError } from 'apollo-server';
import { verify } from 'jsonwebtoken';

const authenticationHandler = (header: string, secret: string): string => {
  const authError = new AuthenticationError('Log in to perform this action');
  if (!header) {
    throw authError;
  }
  const token = header.split(' ')[1];
  // @ts-ignore
  const { userId } = verify(token, secret);
  if (!userId) {
    throw authError;
  }

  return userId;
};
export default authenticationHandler;
