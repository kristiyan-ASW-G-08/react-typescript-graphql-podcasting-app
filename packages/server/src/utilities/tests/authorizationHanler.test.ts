import jwt from 'jsonwebtoken';
import mongoose from 'mongoose';
import authorizationHandler from '../authorizationHandler';

jest.spyOn(jwt, 'verify');
describe('authorizationHandler', () => {
  const secret = 'secret';
  const userId = mongoose.Types.ObjectId();
  const token = jwt.sign(
    {
      userId,
    },
    secret,
    { expiresIn: '1h' },
  );
  const authorizationHeader = `bearer ${token}`;
  it('returns user id', () => {
    expect.assertions(4);
    const returnedUserId = authorizationHandler(authorizationHeader, secret);
    expect(jwt.verify).toHaveBeenCalledTimes(1);
    expect(jwt.verify).toHaveBeenCalledWith(token, secret);
    expect(jwt.verify).toHaveReturnedTimes(1);
    expect(returnedUserId).toMatch(userId.toString());
  });
  it('throw an error when the secret is invalid', () => {
    expect.assertions(1);
    expect(() =>
      authorizationHandler(authorizationHeader, 'invalidSecret'),
    ).toThrowErrorMatchingSnapshot();
  });
  it('throw an error when the secret is invalid', () => {
    expect.assertions(1);
    expect(() =>
      authorizationHandler('', secret),
    ).toThrowErrorMatchingSnapshot();
  });
});
