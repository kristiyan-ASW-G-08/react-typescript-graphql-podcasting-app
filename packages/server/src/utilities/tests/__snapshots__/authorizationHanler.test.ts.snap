// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authorizationHandler throw an error when the secret is invalid 1`] = `"invalid signature"`;

exports[`authorizationHandler throw an error when the secret is invalid 2`] = `"Log in to perform this action"`;
